### ISSUE TITLE
Linux Package

### ISSUE THREAD

**Issue by adambom** [forio/julia-studio#69]
An apt-get package for Linux

**Comment by ViralBShah** [comment]
@sebastien-villemot I think it would be great to have a debian package for julia-studio.

**Comment by svillemot** [comment]
@ViralBShah Yes, this is on my todo list!

**Comment by ViralBShah** [comment]
Thanks!

**Comment by andrioni** [comment]
An AUR PKGBUILD would also be great, and I might try to write one and send it to AUR in the next days.

**Comment by andrioni** [comment]
Just went and wrote one, and it's already [available on AUR](https://aur.archlinux.org/packages/julia-studio-git/)!

**Comment by WestleyArgentum** [comment]
Dude, that's fantastic, thanks!

**Comment by svillemot** [comment]
I had a look at the julia-studio codebase (with the aim of creating an official Debian package), but I have some concerns.

My first concern is that julia-studio appears to be a complete fork of qtcreator. However, the number of files modified or added by julia-studio on top of vanilla qtcreator is small compared to qtcreator itself (essentially a plugin if I understand correctly). This means that adding julia-studio in Debian would introduce a huge code duplication, which is a bad thing for several reasons. Have you explored the possibility of simply shipping a qtcreator plugin for julia, instead of forking the whole qtcreator codebase?

Moreover, my understanding is that julia-studio is currently based on a pre-release of qtcreator 2.6.0, and that further updates to qtcreator in the 2.6.x series have not been merged into julia-studio. This means that a lot of bugfixes to qtcreator have not been incorporated into julia-studio. What is the reason for that?

And finally, how do you plan to deal with future major releases of qtcreator (the current one is already 2.7). Do you plan to permanently fork, or do you intend to rebase your patches on top of the new major releases?

**Comment by ViralBShah** [comment]
I agree that having a complete fork of qtcreator customized for julia is not right for packaging and distribution. It would have been great if qtcreator can have a julia plugin.

**Comment by adambom** [comment]
Maybe we should start a separate issue for this.

**Comment by WestleyArgentum** [comment]
We tried pretty hard to be just a plugin at first. Obviously it would be great for many of the reasons you listed / hinted at. There were several major problems.

Frist off, we simply couldn't do many of the things we wanted to. I'm not talking big interface changes and branding (though those were difficult or impossible as well). I'm talking "let's add something to that context menu... oh, wait, that context menu is completely hard coded. In one of the most core plugins." Once you hit a few of these walls you've either got a list of things you've got to duplicate and re-implement (very time consuming) or you just start trying to make the existing stuff better.

Second, there was no documentation beyond simple "let's make an editor that uses a custom highlighter!" stuff. So even when there were exported interfaces for doing things, they took a long time to find, figure out, and were often developed with something very narrow in mind.

Third, there was no good way of setting up language configurations. So even if you did duplicate large swaths of the existing interface with the improvements you wanted, there was no good way to have yours take priority when they needed to. You could more or less make yours take over, but then you're changing interface people use for other languages.

Off of this, sort of a fourth point, their interface was very complicated for our purposes. We showed it -still laden with much of the standard qt creator interface- to people. Scientists, academics, etc. It was way more c++ and build system setup than someone just wanting to hack some Julia needed. I'm not saying the current interface is amazing, but it's not cluttered and pretty straight forward. Edit code in a file or console. Push the green arrow to make it go.

---

Now, all of that said, we need to do a better job of staying up to date. I really want some of those bug fixes and things are constantly getting better on the Qt Creator end. I'm sure some of the things I mentioned above are already woes of the past.

**Comment by adambom** [comment]
That's a great summary, @WestleyArgentum.

We need to think about what we want to do long term with the code base. I'm not a huge fan of the Qt route, given the enormous bloat and convoluted architecture. I think it's served us well to this point, but we may be hitting a ceiling in terms of doing what we want to do.

By the time we get to 1.0, where do we want to be? But then again, this is probably best suited for another thread.
